File: solutions/python/rail-fence-cipher/1/rail_fence_cipher.py
def rail_cipher(input, rails, do_encode):
    max_gap = 2 * rails - 2
    out_idx = 0
    out = [0] * len(input)

    for i in range(rails):
        idx = i
        gap = max_gap - 2 * i

        if i == rails - 1:
            gap = max_gap

        while idx < len(input):
            if do_encode:
                out[out_idx] = input[idx]
            else:
                out[idx] = input[out_idx]

            idx += gap
            out_idx += 1

            if i != 0 and i != rails - 1:
                gap = max_gap - gap

    return "".join(out)


def encode(message, rails):
    return rail_cipher(message, rails, True)


def decode(encoded_message, rails):
    return rail_cipher(encoded_message, rails, False)

File: solutions/c/rail-fence-cipher/1/rail_fence_cipher.c
#include "rail_fence_cipher.h"

#include <stdlib.h>
#include <string.h>

static char* railCipher(char* input, size_t rails, int isEncoding) {
    int length = strlen(input);
    char* output = malloc(sizeof(char) * (length + 1));
    output[length] = '\0';

    int maxGap = 2 * (int)rails - 2;
    int outputIdx = 0;

    for (size_t i = 0; i < rails; i++) {
        int idx = (int)i;
        int curGap = maxGap - 2 * i;

        if (i == rails - 1) {
            curGap = maxGap;
        }

        while (idx < length) {
            if (isEncoding) {
                output[outputIdx] = input[idx];
            } else {
                output[idx] = input[outputIdx];
            }
            idx += curGap;
            outputIdx++;

            if (i != 0 && i != rails - 1) {
                curGap = maxGap - curGap;
            }
        }
    }

    return output;
}

char* encode(char* text, size_t rails) { 
    return railCipher(text, rails, 1); 
}

char* decode(char* ciphertext, size_t rails) {
    return railCipher(ciphertext, rails, 0);
}
